pyogrio/_err.pyx

```diff
@@ -132,49 +159,114 @@ exception_map = {
     13: CPLE_AWSObjectNotFoundError,
     14: CPLE_AWSAccessDeniedError,
     15: CPLE_AWSInvalidCredentialsError,
-    16: CPLE_AWSSignatureDoesNotMatchError
+    16: CPLE_AWSSignatureDoesNotMatchError,
+    17: CPLE_AWSError
+}
+
+cdef dict _CODE_MAP = {
+    0: 'CPLE_None',
+    1: 'CPLE_AppDefined',
+    2: 'CPLE_OutOfMemory',
+    3: 'CPLE_FileIO',
+    4: 'CPLE_OpenFailed',
+    5: 'CPLE_IllegalArg',
+    6: 'CPLE_NotSupported',
+    7: 'CPLE_AssertionFailed',
+    8: 'CPLE_NoWriteAccess',
+    9: 'CPLE_UserInterrupt',
+    10: 'ObjectNull',
+    11: 'CPLE_HttpResponse',
+    12: 'CPLE_AWSBucketNotFound',
+    13: 'CPLE_AWSObjectNotFound',
+    14: 'CPLE_AWSAccessDenied',
+    15: 'CPLE_AWSInvalidCredentials',
+    16: 'CPLE_AWSSignatureDoesNotMatch',
+    17: 'CPLE_AWSError'
 }
 
 
+cdef class GDALErrCtxManager:
+    """A manager for GDAL error handling contexts."""
+
+    def __enter__(self):
+        CPLErrorReset()
+        return self
+
+    def __exit__(self, exc_type=None, exc_val=None, exc_tb=None):
+        cdef int err_type = CPLGetLastErrorType()
+        cdef int err_no = CPLGetLastErrorNo()
+        cdef const char *msg = CPLGetLastErrorMsg()
+        # TODO: warn for err_type 2?
+        if err_type >= 2:
+            raise exception_map[err_no](err_type, err_no, msg)
```

Comment:
We use `exception_map.get(err_no, CPLE_BaseError)` (for the case that there is an err_no we are not aware of, maybe also for future compatibility?), should we use that here as well?

Reply:
Code removed...